### DISCUSSION TITLE
Help using restore nodes inside recursive dynamic node

### DISCUSSION THREAD

**Discussion by cydmium** [L3MON4D3/LuaSnip#473] [Q&A] [answered]
I am attempting to modify the behavior of the [Latex - Endless list](https://github.com/L3MON4D3/LuaSnip/wiki/Cool-Snippets#latex---endless-list) example on the wiki to effectively reverse the order of operations between choosing and jumping. In the wiki example, to make a new entry in the list, you must first jump to an invisible node then change it to a new list entry. My goal is to first make the new list entry then be able to jump to it. The snippet below achieves this functionality, but introduces another issue. If I write some text on one item then attempt to create a new list entry, the text I previously wrote is removed.
```lua
local recursive_list
recursive_list = function()
  return sn(nil, {
    c(1, {
      sn(nil, {t({"", "\t\\item "}), i(1)}),
      sn(nil, {t({"", "\t\\item "}), i(1), d(2, recursive_list, {})})
    })
  })
end

s("ls", {
  t({"\\begin{itemize}"}),
  d(1, recursive_list, {}),
  t({"", "\\end{itemize}"}),
  i(0)
}),
```

I was hoping to use restore nodes to save the text between choices, but it seems I have no idea how to properly use the restore node functionality. My attempt is below, but it errors when attempting to make more than 2 items. Any advice would be appreciated.

```lua
local recursive_list
recursive_list = function()
  return sn(nil, {
    c(1, {
      sn(nil, {t({"", "\t\\item "}), r(1, "test", i(nil))}),
      sn(nil, {t({"", "\t\\item "}), r(1, "test", i(nil)), d(2, recursive_list, {})})
    })
  })
end

s("ls", {
  t({"\\begin{itemize}"}),
  d(1, recursive_list, {}),
  t({"", "\\end{itemize}"}),
  i(0)
}),
```
Here is the error message:
```
E5108: Error executing lua: ...site/pack/packer/start/LuaSnip/lua/luasnip/util/util.lua:234: Expected lua table
stack traceback:
        [C]: in function 'nvim_buf_set_text'
        ...site/pack/packer/start/LuaSnip/lua/luasnip/util/util.lua:234: in function 'put'
        ...ite/pack/packer/start/LuaSnip/lua/luasnip/nodes/node.lua:63: in function 'put_initial'
        .../pack/packer/start/LuaSnip/lua/luasnip/nodes/snippet.lua:670: in function 'put_initial'
        ...k/packer/start/LuaSnip/lua/luasnip/nodes/restoreNode.lua:115: in function 'put_initial'
        .../pack/packer/start/LuaSnip/lua/luasnip/nodes/snippet.lua:670: in function 'put_initial'
        ...ck/packer/start/LuaSnip/lua/luasnip/nodes/choiceNode.lua:245: in function 'no_region_check_wrap'
        ...nvim/site/pack/packer/start/LuaSnip/lua/luasnip/init.lua:327: in function 'change_choice'
        .../.config/nvim/lua/plugin_config/luasnip.lua:50: in function <.../.config/nvim/lua/plugin_config/luasnip.lua:48>
```

**Comment by L3MON4D3** [comment, marked as answer]
Oh, one immediate problem (not completely sure if that is exactly what's going on here, but it could definitely lead to bugs) I can spot is that only one restoreNode with a given key may be visible at any time, so you'd have to somehow generate a different key for each depth.

I don't think that's easy to do, so I'll instead recommend the [dynamicNode with user input](https://github.com/L3MON4D3/LuaSnip/wiki/Misc#dynamicnode-with-user-input), which makes these recursive-choiceNode cases much cleaner and easier to handle (although it requires a bit of code for setup)

**Reply by cydmium** [reply]
Thanks! That looks like exactly what I need. I don't have a working example yet, but I will update once I get something together.
